dbms/src/Storages/KVStore/Decode/RegionTable.cpp

```diff
@@ -401,18 +427,37 @@ void RegionTable::shrinkRegionRange(const Region & region)
     std::lock_guard lock(mutex);
     auto & internal_region = getOrInsertRegion(region);
     internal_region.range_in_table = region.getRange()->rawKeys();
-    internal_region.updateRegionCacheBytes(region.dataSize());
 }
 
-void RegionTable::extendRegionRange(const RegionID region_id, const RegionRangeKeys & region_range_keys)
+void RegionTable::replaceRegion(const RegionPtr & old_region, const RegionPtr & new_region)
 {
-    std::lock_guard lock(mutex);
+    const auto region_range_keys = new_region->getRange();
+    // Extend region range to make sure data won't be removed.
+    extendRegionRange(*new_region, *region_range_keys);
+    if (old_region)
+    {
+        // `old_region` will no longer contribute to the memory of the table.
+        auto keyspace_id = region_range_keys->getKeyspaceID();
+        auto table_id = region_range_keys->getMappedTableID();
+        auto & table = getOrCreateTable(keyspace_id, table_id);
+        old_region->resetRegionTableCtx();
+        if unlikely (!new_region->getRegionTableCtx())
+        {
+            // For most of the cases, the region is prehandled, so the ctx is set at that moment.
+            new_region->setRegionTableCtx(table.ctx);
+        }
```

Comment:
`getOrCreateTable` should be called under lock protection
```suggestion
        std::lock_guard lock(mutex);
        auto & table = getOrCreateTable(keyspace_id, table_id);
        old_region->resetRegionTableCtx();
        if unlikely (!new_region->getRegionTableCtx())
        {
            // For most of the cases, the region is prehandled, so the ctx is set at that moment.
            new_region->setRegionTableCtx(table.ctx);
        }
```

Reply:
Yes, thx